frontend/catalyst/ag_primitives.py

```diff
@@ -102,23 +108,202 @@ def functional_cond():
     set_state(results)
 
 
+def _call_catalyst_for(start, stop, step, body_fn, get_state, enum_start=None, array_iterable=None):
+    """Dispatch to a Catalyst implementation of for loops."""
+
+    @catalyst.for_loop(start, stop, step)
+    def functional_for(i):
+        if enum_start is None and array_iterable is None:
+            # for i in range(..)
+            body_fn(i)
+        elif enum_start is None:
+            # for x in array
+            body_fn(array_iterable[i])
+        else:
+            # for (i, x) in enumerate(array)
+            body_fn((i + enum_start, array_iterable[i]))
+
+        return get_state()
+
+    return functional_for()
+
+
+def _call_python_for(body_fn, get_state, non_array_iterable):
+    """Fallback to a Python implementation of for loops."""
+
+    for elem in non_array_iterable:
+        body_fn(elem)
+
+    return get_state()
+
+
+def for_stmt(
+    iteration_target: Any,
+    _extra_test: Callable[[], bool] | None,
+    body_fn: Callable[[int], None],
+    get_state: Callable[[], Tuple],
+    set_state: Callable[[Tuple], None],
+    _symbol_names: Tuple[str],
+    _opts: dict,
+):
+    """An implementation of the AutoGraph 'for .. in ..' statement. The interface is defined by
+    AutoGraph, here we merely provide an implementation of it in terms of Catalyst primitives."""
+
+    assert _extra_test is None
+
+    # The general approach is to convert as much code as possible into a graph-based form:
+    # - For loops over iterables will attempt a conversion of the iterable to array, and fall back
+    #   to Python otherwise.
+    # - For loops over a Python range will be converted to a native Catalyst for loop. However,
+    #   since the now dynamic iteration variable can cause issues in downstream user code, any
+    #   errors raised during the tracing of the loop body will restart the tracing process using
+    #   a Python loop instead.
+    # - For loops over a Python enumeration use a combination of the above, providing a dynamic
+    #   iteration variable and conversion of the iterable to array. If either fails, a fallback to
+    #   Python is used.
+    # The fallback mechanism for tracing errors will raise a warning as the user may need to be
+    # aware that the graph conversion failed, for instance for lack of converting lists into arrays,
+    # but the conversion from iterable to array will fall back to Python silently.
```

Comment:
What do you mean by this @dime10?

Reply:
So what I mean is that there are two reasons a fallback to Python could have been triggered:
- the iterable provided by the user is not convertible to an array -> this will fallback to a Python loop silently (without a warning), since there isn't a simple fix to make this loop traceable 
- an exception is raised during the tracing of the loop body after conversion -> this will raise a warning to allow users to correct mistakes and allow the conversion to succeed, for example because they forgot to use a list instead of an array

Reply:
Thanks! Could be worth updating the comment with your github comment, makes it easier to understand :)

Reply:
https://github.com/PennyLaneAI/catalyst/pull/258/commits/515dbcfc2704b676dc275d78238494a61ddc74e3